--- /dev/null
+++ spice-0.14.3/server/hacks.h
@@ -0,0 +1,4 @@
+#define VD_AGENT_SBUTTON_MASK (1 << 6)
+#define VD_AGENT_EBUTTON_MASK (1 << 7)
+#define SPICE_MOUSE_BUTTON_MASK_SIDE (1 << 5)
+#define SPICE_MOUSE_BUTTON_MASK_EXTRA (1 << 6)
--- spice-0.14.3.orig/server/inputs-channel.c
+++ spice-0.14.3/server/inputs-channel.c
@@ -27,6 +27,7 @@
 #include <common/generated_server_marshallers.h>
 #include <common/demarshallers.h>
 
+#include "hacks.h"
 #include "spice.h"
 #include "red-common.h"
 #include "reds.h"
@@ -158,7 +159,9 @@ const VDAgentMouseState *inputs_channel_
 #define RED_MOUSE_BUTTON_STATE_TO_AGENT(state)                      \
     (((state & SPICE_MOUSE_BUTTON_MASK_LEFT) ? VD_AGENT_LBUTTON_MASK : 0) |    \
      ((state & SPICE_MOUSE_BUTTON_MASK_MIDDLE) ? VD_AGENT_MBUTTON_MASK : 0) |    \
-     ((state & SPICE_MOUSE_BUTTON_MASK_RIGHT) ? VD_AGENT_RBUTTON_MASK : 0))
+     ((state & SPICE_MOUSE_BUTTON_MASK_RIGHT) ? VD_AGENT_RBUTTON_MASK : 0) |    \
+     ((state & SPICE_MOUSE_BUTTON_MASK_SIDE) ? VD_AGENT_SBUTTON_MASK : 0) |    \
+     ((state & SPICE_MOUSE_BUTTON_MASK_EXTRA) ? VD_AGENT_EBUTTON_MASK : 0))
 
 static void activate_modifiers_watch(InputsChannel *inputs)
 {
--- spice-0.14.3.orig/server/reds.c
+++ spice-0.14.3/server/reds.c
@@ -53,6 +53,7 @@
 #include <common/generated_server_marshallers.h>
 #include <common/ring.h>
 
+#include "hacks.h"
 #include "spice.h"
 #include "reds.h"
 #include "agent-msg-filter.h"
@@ -1936,7 +1937,9 @@ static void reds_handle_main_link(RedsSt
 #define RED_MOUSE_BUTTON_STATE_TO_AGENT(state)                      \
     (((state & SPICE_MOUSE_BUTTON_MASK_LEFT) ? VD_AGENT_LBUTTON_MASK : 0) |    \
      ((state & SPICE_MOUSE_BUTTON_MASK_MIDDLE) ? VD_AGENT_MBUTTON_MASK : 0) |    \
-     ((state & SPICE_MOUSE_BUTTON_MASK_RIGHT) ? VD_AGENT_RBUTTON_MASK : 0))
+     ((state & SPICE_MOUSE_BUTTON_MASK_RIGHT) ? VD_AGENT_RBUTTON_MASK : 0) |    \
+     ((state & SPICE_MOUSE_BUTTON_MASK_SIDE) ? VD_AGENT_SBUTTON_MASK : 0) |    \
+     ((state & SPICE_MOUSE_BUTTON_MASK_EXTRA) ? VD_AGENT_EBUTTON_MASK : 0))
 
 static void openssl_init(RedLinkInfo *link)
 {
