--- virt-viewer-7.0.orig/src/virt-viewer-timed-revealer.c
+++ virt-viewer-7.0/src/virt-viewer-timed-revealer.c
@@ -177,7 +177,7 @@ virt_viewer_timed_revealer_new(GtkWidget
      * the hidden toolbar.
      */
 
-    gtk_container_add(GTK_CONTAINER(self), priv->revealer);
+//    gtk_container_add(GTK_CONTAINER(self), priv->revealer);
     gtk_widget_set_halign(GTK_WIDGET(self), GTK_ALIGN_CENTER);
     gtk_widget_set_valign(GTK_WIDGET(self), GTK_ALIGN_START);
     gtk_widget_show_all(GTK_WIDGET(self));
